pkg/apiserver/rest/dualwriter.go

```diff
@@ -117,3 +122,50 @@ func (u *updateWrapper) Preconditions() *metav1.Preconditions {
 func (u *updateWrapper) UpdatedObject(ctx context.Context, oldObj runtime.Object) (newObj runtime.Object, err error) {
 	return u.updated, nil
 }
+
+func SetDualWritingMode(
+	kvs *kvstore.NamespacedKVStore,
+	features featuremgmt.FeatureToggles,
+	entity string,
+	legacy LegacyStorage,
+	storage Storage,
+) (DualWriter, error) {
+	toMode := map[string]DualWriterMode{
+		"1": Mode1,
+		"2": Mode2,
+		"3": Mode3,
+		"4": Mode4,
+	}
+	errDualWriterSetCurrentMode := errors.New("failed to set current dual writing mode")
+
+	// Use entity name as key
+	m, ok, err := kvs.Get(context.Background(), entity)
+	if err != nil {
+		return nil, errors.New("failed to fetch current dual writing mode")
+	}
+	if !ok {
+		// Default to mode 1
+		m = fmt.Sprint(Mode1)
+		err := kvs.Set(context.Background(), entity, m)
+		if err != nil {
+			return nil, errDualWriterSetCurrentMode
+		}
+	}
+
+	currentMode := toMode[m]
+
+	// Desired mode is 2 and current mode is 1
+	if features.IsEnabledGlobally(featuremgmt.FlagDualWritePlaylistsMode2) && (currentMode == Mode1) {
+		// This is where we go through the different gates to allow the instance to migrate from mode 1 to mode 2.
```

Comment:
This is a great start!

We're going to have to think about the implications of this being a potentially-blocking process at startup for the switch to mode 3, in theory at least we'd want to start up in mode 2, sync in the background, then switch on the fly to mode 3. That said, it might be simpler to run the sync while we're in mode 2 and just check that we synced successfully (which would be fast) before switching to mode 3.